### PR TITLE
[WIP] Keybinding Overhaul

### PR DESCRIPTION
**Pull request by calmilamsy**
## Summary

**DEPENDS ON #797**
- Add the ability to remap hotkeys.
- Adds in a dynamically generated keybind menu, replacing the old one.

## Pictures/Videos

(First prototype) 
https://cdn.discordapp.com/attachments/603622883075293214/879812051550216283/2021-08-24_20-36-35.mp4

## Changes to Files

- Deletes a bunch of things that are no longer required from the ServerLobby scene.
- Adds in a bunch of logic into GeneralSettingsUIHelper.cs to generate the new keybind menu.
- Changes the BindingButton prefab to have a script attached that rebinds the contained key.
- Changes every script that is ***CURRENTLY*** used in game that I could find that handles keypresses to use the new input system.

## Technical Notes

This one will be a long one once I am done, and I will detail way more once what people should do for keybinds is more set in stone.

- Uses the new Unity Input System (1.1.0-pre.6, requires #797).
- Adds in a helpful public static class `SS3D.Input.InputHelper` that contains all the keybinds which can be accessed from anywhere.
- Adds the ability to dynamically add and remove hotkeys really easily. The keybindings UI currently does NOT support these yet though.

## Known Issues

- Rebinding the scroll wheel for zooming in and out using the keybinds UI is not supported.
- Might be an idea to provide a bit of visual feedback to the user about what key they are rebinding.
- Renaming or removing controls in the InputActions object will cause the binding overrides to implode and throw errors. I have resorted to just letting it stay on default bindings if it fails to load the bindings, overwriting the old bindings config with any new bindings made.

**All of this can change as I make more progress, this is just so folk know I am working on it, and can provide feedback.**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,6 +29,9 @@
 .vs/
 .vscode/
 
+# Visual studio project config
+.vsconfig
```
Comment (cosmiccoincidence) [inline .gitignore:33]: I just added this in a commit on one of Ryan's PRs that was just merged. You can remove this file from your PR.

**Comment by Ryan089** [comment]
This looks really cool. Thanks for the effort you've put into it. Just wanted to let you know it'll probably take quite a while to finish the review, but rest assured it is being looked at!

**Comment by cosmiccoincidence** [comment]
> Rebinding the scroll wheel for zooming in and out using the keybinds UI is not supported.

This is fine because the controls for the camera we have will not be available to players in our final design. These camera controls will be limited to admins only (along with some other controls). But we don't really have that distinguished as of yet.

> Might be an idea to provide a bit of visual feedback to the user about what key they are rebinding.

Did you have an idea in mind on how to do this?

File: Documents/ADDING_CONTROLS.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+5. Done! You can now reference your keybinding using `SS3D.Input.InputHelper.inputs.mycategory.myaction`!
+
+## Implementation Notes (IMPORTANT)
+
+- `myaction.triggered` only runs on the frame a key is pressed and/or released, depending on your press type.
+- `myaction.IsPressed()` will return true if the user is currently holding the key down.
+
+## Technical Notes
+
+- If you set your action to hold, triggered and acts in the exact same way as if you set the press type to Press and Release.
```
Comment (cosmiccoincidence) [inline Documents/ADDING_CONTROLS.md:22]: Typo on this line?

**Review by cosmiccoincidence — COMMENTED** [review]
Some more commenting of what certain code or files do would be appreciated. This will tremendously help new developers in the future get a quicker grasp of what's going on.
A few files I noticed that could use a summary:
- InputHelper.cs
- KeybindingHelper.cs
- KeybindingManager.cs

File: Documents/ADDING_CONTROLS.md (hunk 2 above)
Comment (calmilamsy) [reply Documents/ADDING_CONTROLS.md:22]: Yup, added an extra `and`, woops.

File: .gitignore (hunk 1 above)
Comment (calmilamsy) [reply .gitignore:38]: Removed in latest push.

**Comment by calmilamsy** [comment]
Will work on the code comments and button highlighting on pressing the rebind button tomorrow.

**Comment by calmilamsy** [comment]
> 
> 
> Some more commenting of what certain code or files do would be appreciated. This will tremendously help new developers in the future get a quicker grasp of what's going on.
> A few files I noticed that could use a summary:
> 
>     * InputHelper.cs
> 
>     * KeybindingHelper.cs
> 
>     * KeybindingManager.cs

KeybindingHelper and KeybindingManager were holdovers from the old system. Removed them.
I have also added summaries to everything I added, I think.

**Review by cosmiccoincidence — COMMENTED** [review]
- Are these the defaults? Some of these do not seem correct if they are meant to be the defaults.
image

- I'm a bit confused by this section here. Not sure if we should allow rebinding ESC (the exit key).
image

- There is a scroll bar in the settings menu, but I need to use my mouse to use it. I should also be able to use my scroll wheel on my mouse for this.

- I can't remember, were you going to add a way to reset controls to the default in this PR or are you waiting to add that in another PR?

**Comment by calmilamsy** [comment]
Pretty sure I coped the defaults from the old settings,  I will double check just in case.
I can blacklist escape from being rebound.
Mouse wheel worked in testing, unsure why that isnt working for you.
And I am fairly certain I said I would add resetting to default in another PR, cause it would require rewriting and adding a ton of code that isnt really in the scope of this PR.

**Comment by cosmiccoincidence** [comment]
I recloned the branch tested it all again and I'm still getting weird default keybinds and the scroll bar not working with the mouse wheel.

Also don't forget to disable the ability to rebind the misc category.

Lastly, something new I noticed is that the intent button recently got an update to its appearance but that doesn't seem to be present in your branch here. Consider merging in master to amend that.

**Comment by calmilamsy** [comment]
I *think* I have done all you requested. Please double check to make sure.

**Comment by calmilamsy** [comment]
Moved back to WIP cause I am changing almost everything about the Keybind UI.

**Comment by cosmiccoincidence** [comment]
Closing this for now and marking it Derelict so we come back to it.

**Pull request closed without merging**